### ISSUE TITLE
[IMPROVEMENT] Add a label and selector for webhooks to longhorn-manager pod manifest.

### ISSUE THREAD

**Issue by james-munson** [longhorn/longhorn#8803]
## Is your improvement request related to a feature? Please describe (👍 if you like this request)

This came up in a discussion of the testing for #6205 where sometimes changes requested for fast failover, such as modifying the volume attachment, would time out because the admission webhook that was asked to validate the change may have been on the node that was failed but is not yet marked as "not ready."  In that case, the webhook request times out and will do so until the dead node is removed from the service endpoint slice by Kubernetes.

## Describe the solution you'd like

@PhanLe1010 suggested this change, which adds a label to longhorn-manager pod for each webhook,
```yaml
  longhorn.io/component: admission-webhook 
  longhorn.io/component: conversion-webhook
```
and a selector on that label in the respective services.

If a node is suspected dead, the label can be removed by another node, and the dead node's IP will be removed from the endpoint slice.  When the node returns, it can add the label back to its own longhorn-manager pod's manifest when it is ready for the webhook to become active.

Although this necessary for reliable fast failover, it can also be used to control when the pod is considered ready, since the readiness probe uses the webhook to answer for the pod.

**Comment by PhanLe1010** [comment]
> Although this necessary for reliable fast failover, it can also be used to control when the pod is considered ready, since the readiness probe uses the webhook to answer for the pod.

Could you give a bit more details about this point. Now thinking about it, I am not sure how we can achieve this readiness probe effect

**Comment by ejweber** [comment]
> Could you give a bit more details about this point. Now thinking about it, I am not sure how we can achieve this readiness probe effect

IIRC, we weren't discussing this in conjunction with the readiness probe per se. I thought the idea was:

- The readiness probe can pass as soon as the conversion webhook is up.
- The admission webhook may not be up at that time.
- We don't want a pod with only the conversion webhook ready to receive admission webhook requests. They will fail.
- The pod can label itself when it is ready for either type of traffic (individually), so each appropriate service will only direct traffic to pods ready for that type of traffic.

**Comment by derekbit** [comment]
> When the node returns, it can add the label back to its own longhorn-manager pod's manifest when it is ready for the webhook to become active.

Could you give more details about the design and implementation?  
- Which controller is responsible for removing and adding the labels?
- How to check if a node is dead?

**Comment by innobead** [comment]
@james-munson the milestone was missed. If working on it, need to ensure the milestone is set as well.

**Comment by innobead** [comment]
> ## Is your improvement request related to a feature? Please describe (👍 if you like this request)
> This came up in a discussion of the testing for #6205 where sometimes changes requested for fast failover, such as modifying the volume attachment, would time out because the admission webhook that was asked to validate the change may have been on the node that was failed but is not yet marked as "not ready." In that case, the webhook request times out and will do so until the dead node is removed from the service endpoint slice by Kubernetes.

The admission webhook is based on the `service`, and each Longhorn manager pod runs with the webhook. If one is unable to process the request, then the request should be handled by the others. Why does it get stuck until the dead node is removed? Or anything I missed?

**Comment by ejweber** [comment]
> The admission webhook is based on the service, and each Longhorn manager pod runs with the webhook. If one is unable to process the request, then the request should be handled by the others. Why does it get stuck until the dead node is removed? Or anything I missed?

Once Kubernetes realizes the node a longhorn-manager pod is running on is not ready, it no longer directs traffic to the associated endpoint (regardless of the pod status). For example, in my cluster, with a node down, I see:

```
# Pod status
longhorn-manager-nlm8h   1/1     Running   4 (4d2h ago)   6d18h

# Nodes status
eweber-v126-worker-9c1451b4-kgxdq   NotReady   worker                      176d   v1.28.9+k3s1

# Part of EndpointSlice status
- addresses:
  - 10.42.2.36
  conditions:
    ready: false
    serving: false
    terminating: false
  nodeName: eweber-v126-worker-9c1451b4-kgxdq
  targetRef:
    kind: Pod
    name: longhorn-manager-nlm8h
    namespace: longhorn-system
    uid: 4333ddfb-4b16-4f8b-a284-771bff29cab9
```

So the node is not ready and traffic is not routed to the pod (which is supposedly running, but really Kubernetes just doesn't know its situation since kubelet is down).

@james-munson has the logs, but we saw:

- Traffic is still routed to the pod on the down node until the node is NotReady.
- For every API request involving a Longhorn object, there is a API server -> conversion webhook request and an API server -> admission webhook request.
- Each request has a 10 second timeout.
- The requests are routed to each backing pod (including the down one) randomly (and/or in round robin style).

So, in a three node cluster with one node down node, a single request to the API server has:

- A 66.6% * 66.6% = 44.4% chance of reaching a functional pod on the first try.
- A 33.3% chance of reaching a down pod for the conversion webhook and stalling for 10 seconds before failing.
- A 66.6% * 33.3% = 22.2% chance of reaching a functional pod for the conversion webhook but a down pod for the admission webhook and stalling for 10 seconds before failing.

Now cumulatively:

- In total, that's a 55.5% chance of AT LEAST a 10 second delay (and a 55.5% * 44.4% = 30.8% chance of ONLY a 10 second delay).
- Then, there's a 55.5% * 55.5% = 30.8% chance of AT LEAST a 20 second delay (and a 55.5% * 44.4% = 24.6% chance of ONLY a 20 second delay).

Wrapping it up, that's a:

- 44.4% chance of no delay
- 24.6% chance of a 10 second delay
- 13.7% chance of a 20 second delay
- 7.6% chance of a 30 second delay
- 4.2% chance of a 40 second delay
- 5.3% chance of an even longer delay

If multiple requests are required for an action to complete (e.g. related to two different resources), the chances of getting stuck increase.

TLDR: You are right that there are other pods to service the webhook requests, but we may still be delayed for a really long time trying to get through to a working one. If it takes longer than ~40 seconds to do so, then @james-munson's fast failover work becomes relatively meaningless.

**Comment by ejweber** [comment]
I was thinking of another idea @james-munson and @PhanLe1010. I can't remember if I said it during the workshop or not.

In the fast failover case, we could choose to delete the longhorn-manager pod on the node we suspect is down instead of using this label mechanism. This should also stop traffic from routing to the pod (which will be terminating).

On the one hand, it is probably MUCH less disruptive than deleting the instance-manager pod (or otherwise considering all engines to be crashed) in the case that the node is not actually down (false positive). On the other hand, it will cause a bunch of resource ownership changes.

**Comment by james-munson** [comment]
Thank you @ejweber for the detailed explanation of the issue.  That hits the nail on the head.

We did discuss the possibility of simply deleting the longhorn-manager pod on the suspected node.  If the suspicion is false and the node is actually up,  a new pod is created within seconds.  I'm not sure that there will be resource changes, since those also depend (I believe) on the node's ready state, not the existence of a pod.  Those resources owned by the node will go un-synced for a few seconds, but no other node would take responsibility.

Simple as that solution might be, we did hesitate to go that far, and the label/selector solution was attractigve because it addresses #8612 neatly.

**Comment by james-munson** [comment]
As @ejweber also pointed out, if the labels are added dynamically, they should not be in the manifest yaml.  Only the selectors should be.  There will need to be a separate change in longhorn-manager to add the labels at startup, which should go here in the daemon startManager function https://github.com/longhorn/longhorn-manager/blob/1c19bfbd35d8fd59dba136cfc12a879c91b68f7c/app/daemon.go#L162 between `StartWebhook` and `CheckWebhookServiceAvailability` for the conversion webhook, and similarly for the admission webhook.

That should be sufficient for this ticket and for #8612.  For HA, it is necessary to know that the node is probably dead, which will use the `Delinquent` condition introduced in #6205.  It will have an additional change to remove the label when the condition is set to true.  If the node is not actually dead, the labels can be restored at the same time the condition is cleared, in the node_handler's sync code.  But that additional modification will be done in the other feature branch after this change is pulled in.

**Comment by PhanLe1010** [comment]
Lot of good statistics at https://github.com/longhorn/longhorn/issues/8803#issuecomment-2197297624 @ejweber 🚀 

:D

**Comment by PhanLe1010** [comment]
> I was thinking of another idea @james-munson and @PhanLe1010. I can't remember if I said it during the workshop or not.
> 
> In the fast failover case, we could choose to delete the longhorn-manager pod on the node we suspect is down instead of using this label mechanism. This should also stop traffic from routing to the pod (which will be terminating).
> 
> On the one hand, it is probably MUCH less disruptive than deleting the instance-manager pod (or otherwise considering all engines to be crashed) in the case that the node is not actually down (false positive). On the other hand, it will cause a bunch of resource ownership changes.

I remembered we discussed this in the initial meeting long time ago? Yeah, the concern is the ownership transfer storm. And like @james-munson mentioned, a new Longhorn manager pod might be recreated quickly and prevent the RWX volume from detaching from that node? Though I agree that it will be useful to give this a POC if time allows, WDYT @james-munson ?

**Comment by ejweber** [comment]
> Could you explain a bit why 55.5% * 44.4% = 24.6% chance of ONLY a 10 second delay ?

Yeah, I had to calculate this all out a few times before coming to results I was confident in.

- Every time we make an attempt to update a CR, there is a 44.4% chance of failure and a 55.5% chance of success.
- There is a 55.5% chance of a failure on the first attempt of a CR update.
- GIVEN that this occurred (i.e. GIVEN the 55.5% "path"), there is a 44.4% chance of success on the next attempt. This combination of events (55.5% * 44.4%) results in ONLY a ten second delay (caused by the failure on the first attempt). Otherwise, a failure occurs on the second attempt, leading to a further delay.

Here is a visual representation. The bolded probabilities are the "final" probability for a particular delay. There is still a 5.3% chance of a longer delay (so they don't add up to 100%).

![image](https://github.com/longhorn/longhorn/assets/24213029/8ec563c0-4ffc-40e1-8785-b269b0d8345c)

**Comment by PhanLe1010** [comment]
🚀 That is a great statistic calculation! Many thanks!

**Comment by derekbit** [comment]
@james-munson Please update the status of the ticket. Thank you.

**Comment by roger-ryao** [comment]
**Verified on master-head 20240726**
- longhorn `master-head` https://github.com/longhorn/longhorn/commit/d5e8a44104f5e03c310870b47da7c8ab55f671b3

**Verified on v1.7.0-rc2 20240726**
- longhorn `v1.7.0-rc2` https://github.com/longhorn/longhorn/commit/03d5faee809d984a71a17c04da863d571df009bb

**The test steps**
https://github.com/longhorn/longhorn/issues/8803#issuecomment-2194730868

1. List `Longhorn-mamanger` IP : Use `kubectl get pod -n longhorn-system -owide |grep longhorn-manager` to list all `Longhorn-mamanger` IPs in the `longhorn-system` namespace.

2. Describe each service to check selectors
```bash
kubectl describe svc longhorn-admission-webhook -n longhorn-system |grep Selector
kubectl describe svc longhorn-conversion-webhook -n longhorn-system |grep Selector
kubectl describe svc longhorn-recovery-backend -n longhorn-system |grep Selector
```

3. Describe each endpoint to check IP addresses
```bash
kubectl describe endpoints longhorn-admission-webhook -n longhorn-system |grep Addresses
kubectl describe endpoints longhorn-conversion-webhook -n longhorn-system |grep Addresses
kubectl describe endpoints longhorn-recovery-backend -n longhorn-system |grep Addresses
```
**Result Passed**


The test passed under the following scenarios:

1. Fresh install of `v1.7.0-rc2`
2. Fresh install of `master-head`
3. Upgrade from `v1.6.2` to `v1.7.0-rc2`
